src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala

```diff
@@ -0,0 +1,181 @@
+package mesosphere.marathon.core.task.termination.impl
+
+import akka.actor.{ Actor, ActorLogging, Cancellable, Props }
+import mesosphere.marathon.MarathonSchedulerDriverHolder
+import mesosphere.marathon.core.base.Clock
+import mesosphere.marathon.core.task.termination.TaskKillConfig
+import mesosphere.marathon.state.Timestamp
+import mesosphere.marathon.core.task.{ Task, TaskStateOp }
+import mesosphere.marathon.core.task.tracker.{ TaskStateOpProcessor, TaskTracker }
+import mesosphere.marathon.core.event.MesosStatusUpdateEvent
+
+import scala.collection.mutable
+import scala.concurrent.Promise
+
+/**
+  * An actor that handles killing tasks in chunks and depending on the task state.
+  * Lost tasks will simply be expunged from state, while active tasks will be killed
+  * via the scheduler driver. There is be a maximum number of kills in flight, and
+  * the service will only issue more kills when tasks are reported terminal.
+  */
+private[impl] class TaskKillServiceActor(
+    taskTracker: TaskTracker,
+    driverHolder: MarathonSchedulerDriverHolder,
+    stateOpProcessor: TaskStateOpProcessor,
+    config: TaskKillConfig,
+    clock: Clock) extends Actor with ActorLogging {
+  import TaskKillServiceActor._
+  import context.dispatcher
+
+  val tasksToKill: mutable.HashMap[Task.Id, Option[Task]] = mutable.HashMap.empty
+  val inFlight: mutable.HashMap[Task.Id, TaskToKill] = mutable.HashMap.empty
+  var retryTimer: Option[Cancellable] = None
+
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[MesosStatusUpdateEvent])
+  }
+
+  //scalastyle:off cyclomatic.complexity
+  override def receive: Receive = {
+    case KillTaskById(taskId) =>
+      killTaskById(taskId)
+
+    case KillTask(task) =>
+      killTask(task)
+
+    case KillUnknownTaskById(taskId) =>
+      killUnknownTaskById(taskId)
+
+    case KillTasks(tasks, promise) =>
+      killTasks(tasks, promise)
+
+    case Terminal(event) if inFlight.contains(event.taskId) || tasksToKill.contains(event.taskId) =>
+      handleTerminal(event.taskId)
+
+    case Retry =>
+      retry()
+
+    case unhandled: Request =>
+      log.warning("Received unhandled {}", unhandled)
+
+    case unexpected: Any =>
+      log.debug("Received unexpected {}", unexpected)
+  }
+  //scalastyle:on
+
+  def killTaskById(taskId: Task.Id): Unit = {
+    log.debug("Received KillTaskById({})", taskId)
+    import context.dispatcher
+    taskTracker.task(taskId).map {
+      case Some(task) => self ! KillTask(task)
+      case _ => self ! KillUnknownTaskById(taskId)
+    }
+  }
+
+  def killTask(task: Task): Unit = {
+    log.debug("Received KillTask({})", task)
+    tasksToKill.update(task.taskId, Some(task))
```

Comment:
To prevent shared mutable state and separate concerns, but most of all because there were observed kill sequences before as well as fire and forget killings (this sounds way too figurative ...). A `TaskKillProgressActor`'s data will usually be a subset of the `TaskKillServiceActor`'s set. Killing tasks does not imply the creation of a `TaskKillProgressActor`, as not all users of this service wait for a future. When, for example, 10 apps are deleted, there will be 10 progress actors watching for the respective tasks to terminate, 1 for each `killTasks(...)` call. Yet there might be an Upgrade of another app that simply chimes in several kill requests and doesn't wait for anything but listen to the eventStream. As noted in a comment above, that might be adjusted in the future.

So, overall,
* The `TaskKillServiceActor`'s concern is to kill a given set of tasks, throttle and retry (both of these _could_ be separated if necessary), while
* The `TaskKillProgressActor`'s concern is to watch a given set of tasks and complete a promise when all of them are reported terminal.

I acknowledge there is or should be a strong relation between those concerns, and I'm open to refactor this when we tackle the deployment logic (see `TaskReplaceActor` for an example).